### ISSUE TITLE
Making lowliner better understand data frames: by_group()

### ISSUE THREAD

**Issue by lionel-** [hadley/purrr#43]
This implements a functional that takes a grouped table and feed each group to `.f`. It returns a data frame and handles both mutating and summarising operations, the only constraint being that the number of returned rows is the same inside each group. The grouping variables are recycled to match the output size. If `.f` does not return a data frame, `by_group()` returns a list-column.

I also wrote `set_groups()` to set the grouping attributes because it looks a bit funny to have
```{r}
  df %>%
    group_by(col) %>%
    by_group(fun)
```
and also this way we have a non-NSE version of `group_by()` that handles vectors of names and column positions.

In action:
```{r}
# Action operation
mtcars %>%
  set_groups("cyl") %>%
  by_group(partial(lm, mpg ~ disp))
```
```{r}
## + + Source: local data frame [3 x 2]
## 
##   cyl     out
## 1   4 <S3:lm>
## 2   6 <S3:lm>
## 3   8 <S3:#lm>
```

```{r}
# Summarizing operation
mtcars %>%
  set_groups(c("cyl", "am")) %>%
  by_group(map, mean)
```
```{r}
## + + Source: local data frame [6 x 11]
## 
##   cyl am    mpg    disp      hp   drat     wt   qsec    vs   gear   carb
## 1   4  0 22.900 135.867  84.667 3.7700 2.9350 20.970 1.000 3.6667 1.6667
## 2   4  1 28.075  93.612  81.875 4.1837 2.0423 18.450 0.875 4.2500 1.5000
## 3   6  0 19.125 204.550 115.250 3.4200 3.3887 19.215 1.000 3.5000 2.5000
## 4   6  1 20.567 155.000 131.667 3.8067 2.7550 16.327 0.000 4.3333 4.6667
## 5   8  0 15.050 357.617 194.167 3.1208 4.1041 17.143 0.000 3.0000 3.0833
## 6   8  1 15.400 326.000 299.500 3.8800 3.3700 14.550 0.000 5.0000 6.0000
```

```{r}
# Muutating operation
mtcars %>%
  set_groups(c("cyl", "am")) %>%
  by_group(map, ~ . / sd(.))
```
```{r}
## + + Source: local data frame [32 x 11]
## 
##    cyl am     mpg    disp     hp   drat     wt   qsec     vs    gear
## 1    4  0 16.7977 10.5015 3.1544 28.385 7.8278 11.966    Inf  6.9282
## 2    4  0 15.6962 10.0792 4.8333 30.154 7.7296 13.701    Inf  6.9282
## 3    4  0 14.8012  8.5974 4.9350 28.462 6.0487 11.972    Inf  5.1962
## 4    4  1  5.0849  5.2743 4.1050 10.572 5.6675 16.538 2.8284  8.6410
## 5    4  1  7.2259  3.8434 2.9132 11.203 5.3744 17.303 2.8284  8.6410
## 6    4  1  6.7799  3.6969 2.2953 13.537 3.9453 16.458 2.8284  8.6410
## 7    4  1  7.5605  3.4722 2.8691 11.588 4.4827 17.685 2.8284  8.6410
## 8    4  1  6.0885  3.8580 2.9132 11.203 4.7270 16.796 2.8284  8.6410
## 9    4  1  5.7986  5.8749 4.0167 12.164 5.2278 14.841 0.0000 10.8012
## 10   4  1  6.7799  4.6443 4.9878 10.352 3.6961 15.019 2.8284 10.8012
.. ... ..     ...     ...    ...    ...    ...    ...    ...     ...
```

**Comment by lionel-** [comment]
A first implementation is in https://gist.github.com/lionel-/0b41c6b9d3554725807a

**Comment by hadley** [comment]
I'd really prefer to keep row-based data frame stuff in dplyr

**Comment by lionel-** [comment]
I think that there is a need for row-based stuff in lowliner because there are many problems that are a bit  difficult or awkward to solve in dplyr due to NSE.

Also dplyr's DSL interface and the FP mapping functions of lowliner are very different, so there is a rather dire mental switch when we go from `map(df, .f)` to `do()` and the `manip_each(df, funs())` variants... Plus the distinction between summarising and mutating operations is not so relevant when it comes to mapping a function to each column.

I love dplyr and tidyr for cleaning and reshaping a data frame, but when it comes to do actual work with the cleaned up data, it's also nice to have a purely functional interface.

**Comment by hadley** [comment]
In that case the challenge is coming up with a naming convention for lowliner that clearly distinguishes it from dplyr so there no confusion about what
Interface you're using. The documentation will need to include the caveat that dplyr will be much much faster for most scenarios given its Rcpp backend.

**Comment by lionel-** [comment]
It's true that `by_group()` is close to `group_by()` so I tried hard to find another nice name but it's indeed challenging. But after using it a bit, I didn't find it that confusing if I don't use them in the same block of code.

It's a bit kaleidoscopic-looking to `group_by(col)` and then `by_group(fun)` in one go, and also not very nice to mix a NSE and non NSE interface in one pipe chain, so the preferred way to group a data frame before a grouped action would be with the new `set_groups()` function.

Do you think `by_group()` is too confusing?

**Comment by hadley** [comment]
Yeah, I think `by_group()` is too confusing - there needs to be a clear semantic distinction between dplyr and lowliner.

**Comment by lionel-** [comment]
Maybe:
```
mtcars %>%
  set_slices(c("am", "cyl")) %>%
  by_slice(map, mean)
```

This would be good for culinary piping examples : `pizza %>% by_slice(eat)`

**Comment by hadley** [comment]
How about `slice_df()`? Or `slice_rows()`? A common suffix would help make the distinction clear.

**Comment by lionel-** [comment]
I like `slice_rows()` a lot.

I'll be away for a week, I'll make a PR after that.
I'll also try to implement the subsetting in Rcpp, it should be a good project to learn some c++

**Comment by lionel-** [comment]
@hadley,

In case of incompatible data frame outputs in each slice, do you prefer that `by_slice()` fails with an error (through `bind_rows()`) or is it ok to downcast the output to a list-column of data frames?

I like downcasting but I have a feeling you prefer a failure, in the same way that `map_if()` fails if the outputs are not compatible. And it would still be possible to wrap the outputs manually by specifying `..f = . %>% fun() %>% list()`.

**Comment by hadley** [comment]
You got it - I'd definitely prefer a failure

**Comment by lionel-** [comment]
So I have a couple of Rcpp implementations. They are not finished yet but I have some weird performance results.

The first implementation calls `..f` from Rcpp while the second one only uses Rcpp for subsetting the data frame (using `dplyr::DataFrameVisitors` c++ class) and applies the function from R. I did some benchmarks that I can't make sense of.

In the first test I apply a linear model to each group. The first Rcpp implementation is slow, which I expected since there is some overhead to calling a R function from Rcpp.

```{r}
data <- rerun(1000, mtcars) %>% bind_rows() %>% group_by(cyl, vs)

fnu <- partial(lm, disp ~ gear)
microbenchmark(
  R = data %>% by_slice(fnu),
  dplyr = data %>% do(as_data_frame(list(.out = list(fnu(.))))),
  Rcpp1 = data %>% by_slice2(fnu),
  Rcpp2 = data %>% by_slice3(fnu)
)
## Unit: milliseconds
##    expr     min      lq  median      uq    max neval
##       R  23.622  27.051  27.925  29.245 136.09   100
##   dplyr  22.745  25.326  26.603  28.003 132.03   100
##   Rcpp1 425.285 429.163 430.672 434.809 535.71   100
##   Rcpp2  20.345  22.702  23.748  24.961 129.81   100
```

Then in the second test I get the group averages of each column. Now the first Rcpp version is faster, almost on par with dplyr.

```{r}
fnu2 <- partial(map, .f = mean)
microbenchmark(
  R = data %>% by_slice(fnu2),
  dplyr = data %>% summarise_each(funs(mean)),
  Rcpp1 = data %>% by_slice2(fnu2),
  Rcpp2 = data %>% by_slice3(fnu2)
)
## Unit: milliseconds
##    expr    min     lq median     uq      max neval
##       R 5.7227 6.6822 7.2341 8.9383 103.5715   100
##   dplyr 1.9832 2.1714 2.2717 2.4594   5.1598   100
##   Rcpp1 2.3717 2.6066 2.8877 3.3270   7.0950   100
##   Rcpp2 2.8226 3.0679 3.2785 3.8020  96.7852   100
```

The code is in https://gist.github.com/lionel-/0b41c6b9d3554725807a#file-by-slice-cpp. I don't see any warning or error from `lm()` so I don't think the slowness in the first test is caused by exception catching...

I think I'll go with the second implementation (I still need to replicate the ID columns, I'll borrow some code from reshape for that) but I post this here in case you have an idea of what's going on.

**Comment by lionel-** [comment]
Here are a few more benchmarks. The first Rcpp implementation (the one that evaluates `..f` from Rcpp) gets better when the groups are smaller. So I think it has to do with the size of the lm objects. Maybe Rcpp has trouble handling large objects in List (the return type), or passing them to R. To check that, I ran the model with `big_lm()` and indeed the perf gap disappears. Other than that, it's faster to call `..f` in Rcpp. Not sure how to work around this.

*linear model, 32000 rows, 5 groups:*

```{r}
mb_few_groups_lm %>% print(unit = "relative", digits = 2)
##          expr median     unit
## 1           R   1.27 relative
## 2       dplyr   1.27 relative
## 3       Rcpp1  19.56 relative
## 4 Rcpp1_biglm   1.00 relative
## 5       Rcpp2   1.11 relative
```

*linear model, 32000 rows, 1000 groups:*

```{r}
mb_many_groups_lm %>% print(unit = "relative", digits = 2)
##          expr median         unit
## 1           R   1.64 relative
## 2       dplyr   1.68 relative
## 3       Rcpp1   2.35 relative
## 4 Rcpp1_biglm   1.00 relative
## 5       Rcpp2   1.59 relative
```


In the following tests I'm taking group averages using mean. I have two dplyr's versions, one that relies on hybrid evaluation, and the second that does not.

*group average, 32000 rows, 5 groups:*

```{r}
mb_few_groups_mean %>% print(unit = "relative", digits = 2)
## neval: 100
##     expr median     unit
## 1       R   3.62 relative
## 2   dplyr   1.00 relative
## 3  dplyr2   1.33 relative
## 4   Rcpp1   1.56 relative
## 5   Rcpp2   1.71 relative
```

*group average, 32000 rows, 1000 groups:*

```{r}
mb_many_groups_mean %>% print(unit = "relative", digits = 2)
## neval: 100
##     expr median     unit
## 1       R  87.33 relative
## 2   dplyr   1.00 relative
## 3  dplyr2  15.54 relative
## 4   Rcpp1  58.79 relative
## 5   Rcpp2  83.48 relative
```

So judging from the performance of the non-hybrid dplyr version, my Rcpp implementations still have some way to go. Maybe this is because dplyr evaluates a lazy expression instead of applying functions directly? I will try a simple lazy version (just to understand where the perf gap comes from, since you don't want NSE in purrr) once I understand dplyr's code better.

**Comment by lionel-** [comment]
> Maybe Rcpp has trouble handling large objects in List (the return type), or passing them to R.

The problem is that the () operator of Rcpp's Function class creates a language object with the function and its arguments before calling `Rcpp_eval()`. To do that, it expands the argument in the same way that `do.call()` does. So large data frames will create a lot of overhead. To see it:

```{c++}
// [[Rcpp::export]]
SEXP eval_lang1(SEXP& fun, SEXP& arg) {
  return Rcpp_lcons(fun, pairlist(arg));
}

/*** R
eval_lang1(mean, mtcars)
*/
```

```{r}
## (function (x, ...) 
## UseMethod("mean"))(list(mpg = c(21, 21, 22.8, 21.4, 18.7, 18.1, 
## 14.3, 24.4, 22.8, 19.2, 17.8, 16.4, 17.3, 15.2, 10.4, 10.4, 14.7, 
## 32.4, 30.4, 33.9, 21.5, 15.5, 15.2, 13.3, 19.2, 27.3, 26, 30.4,
## ....
```

**Comment by hadley** [comment]
Nice! I think you now know more about the C++ side of dplyr than I do!

**Comment by lionel-** [comment]
aha, probably not :-)

By the way, to expand on my previous edit, it's not the expansion by itself that is costly, but its evaluation by `Rf_eval()`. This cost doesn't occur in `lapply()`'s C implementation but I still don't understand why. Maybe it's because it's a special function that gets its arguments unevaluated.

**Comment by lionel-** [comment]
Alright, traced it back to dots expansion!

It's because I used as mapping function
```
partial(lm, disp ~ cyl)
```
which gets all its arguments through dots. If I use instead
```
function(data) lm(disp ~ cyl, data)
```
there is no performance cost.

```
neval: 100
                 expr median         unit
1    dots,    raw arg   2309 microseconds
2    dots, quoted arg   1038 microseconds
3 no dots,    raw arg   1020 microseconds
4 no dots, quoted arg   1058 microseconds
```

It may be worth it to rewrite `partial()` to return a function with the same filled in signature as the original function.

Anyway, `lapply()` and `map()` are immune to this, and I also use a trick in `by_slice()` to avoid this cost (I put the slices in an environment and construct a call with `..f` and a symbol so there is no language expansion).

**Comment by lionel-** [comment]
Traced it further to `match.call()`, which is run by `lm()`. Apparently, evaluating an inlined call running `match.call()` has no effect on performance, *except* when it needs to deal with dots.

@hadley, on a related matter, did you notice that even in `do.call()`, inlining calls has no performance cost anymore? It seems an R update has fixed this. You may want to update http://rpubs.com/hadley/do-call2. Inlining only has consequences if the function passed to `do.call()` runs `match.call()`, or if an exception is raised.

I wrote a c++ version of `do.call()` that has none of those two problems. Should I send a PR so that `stitch()` and `match.call()` use it? It works by assigning each argument in an environment (with names `_1`, `_2`, etc) before evaluating, so there is no inlining.

It's a few micro-seconds slower than `do.call()`:
```{r}
neval: 100
                                   expr median         unit
1         do.call(nrow, list(diamonds))      7 microseconds
2  do.call(nrow, list(quote(diamonds)))      7 microseconds
3        map_call(list(diamonds), nrow)     10 microseconds
4 map_call(list(quote(diamonds)), nrow)     11 microseconds
```

But it has no perf cost when `match.call()` is used or an exception is raised:
```{r}
                                         expr median         unit
1         do.call(match_call, list(diamonds))   1875 microseconds
2  do.call(match_call, list(quote(diamonds)))      8 microseconds
3        map_call(list(diamonds), match_call)      8 microseconds
4 map_call(list(quote(diamonds)), match_call)      9 microseconds
```

```{r}
system.time(do.call(halt, list(diamonds)))
## Error in (function (x)  (from do-call.cpp!3846PsD#2) : !
## Timing stopped at: 0.67 0.004 0.683 

system.time(map_call(list(diamonds), halt))
## Error in (function (x)  (from do-call.cpp!3846PsD#2) : !
## Timing stopped at: 0 0 0 
```

**Comment by lionel-** [comment]
Closed by #51
